src/codegen/spirv.zig

```diff
@@ -84,23 +385,197 @@ pub const SPIRVModule = struct {
+            .div => try self.genBinOp(inst.castTag(.div).?),
+            .bit_and => try self.genBinOp(inst.castTag(.bit_and).?),
+            .bit_or => try self.genBinOp(inst.castTag(.bit_or).?),
+            .xor => try self.genBinOp(inst.castTag(.xor).?),
+            .cmp_eq => try self.genBinOp(inst.castTag(.cmp_eq).?),
+            .cmp_neq => try self.genBinOp(inst.castTag(.cmp_neq).?),
+            .cmp_gt => try self.genBinOp(inst.castTag(.cmp_gt).?),
+            .cmp_gte => try self.genBinOp(inst.castTag(.cmp_gte).?),
+            .cmp_lt => try self.genBinOp(inst.castTag(.cmp_lt).?),
+            .cmp_lte => try self.genBinOp(inst.castTag(.cmp_lte).?),
+            .bool_and => try self.genBinOp(inst.castTag(.bool_and).?),
+            .bool_or => try self.genBinOp(inst.castTag(.bool_or).?),
+            .not => try self.genUnOp(inst.castTag(.not).?),
+            .arg => self.genArg(),
+            // TODO: Breakpoints won't be supported in SPIR-V, but the compiler seems to insert them
+            // throughout the IR.
+            .breakpoint => null,
+            .dbg_stmt => null,
+            .ret => self.genRet(inst.castTag(.ret).?),
+            .retvoid => self.genRetVoid(),
+            .unreach => self.genUnreach(),
+            else => self.fail(.{.node_offset = 0}, "TODO: SPIR-V backend: implement inst {}", .{inst.tag}),
+        };
+    }
+
+    fn genBinOp(self: *DeclGen, inst: *Inst.BinOp) !u32 {
+        // TODO: Will lhs and rhs have the same type?
+        const lhs_id = try self.resolve(inst.lhs);
+        const rhs_id = try self.resolve(inst.rhs);
+
+        const result_id = self.spv.allocResultId();
+        const result_type_id = try self.getOrGenType(inst.base.ty);
+
+        // TODO: Is the result the same as the argument types?
+        // This is supposed to be the case for SPIR-V.
+        std.debug.assert(inst.rhs.ty.eql(inst.lhs.ty));
+        std.debug.assert(inst.base.ty.tag() == .bool or inst.base.ty.eql(inst.lhs.ty));
+
+        // Binary operations are generally applicable to both scalar and vector operations in SPIR-V, but int and float
+        // versions of operations require different opcodes.
+        // For operations which produce bools, the information of inst.base.ty is not useful, so just pick either operand
+        // instead.
+        const info = try self.arithmeticTypeInfo(inst.lhs.ty);
+
+        if (info.class == .composite_integer)
+            return self.fail(.{.node_offset = 0}, "TODO: SPIR-V backend: binary operations for composite integers", .{});
+
+        const is_bool = info.class == .bool;
+        const is_float = info.class == .float;
+        const is_signed = info.signedness == .signed;
+        // **Note**: All these operations must be valid for vectors of floats, integers and bools as well!
+        // For floating points, we generally want ordered operations (which return false if either operand is nan).
+        const opcode = switch (inst.base.tag) {
```

Comment:
I suggest to do this switch only once (above in `genInst`). I don't think the attempt to avoid code duplication by making this `genBinOp` abstraction is doing you any favors here.

Reply:
I think that like this it keeps all the relevant information quite neatly in the function that actually deals with binary operators. What would you otherwise suggest? I assume moving the `if` conditions up to the genInst, but in that case it would be rather a pain to get the arithmeticTypeInfo of one of the argument types.